### PR TITLE
Make grantToken tenant-aware

### PR DESCRIPTION
**Pull request by lisajian**
### Description

Adds multi-tenancy support on grantToken endpoint by pulling tenant ID from refresh token. Changes include:
- Make refresh token minting / validating stateless by serializing RefreshTokenRecord instead of representing with random string 
- Changing server side logic to check request body's refresh token for tenant ID (note that `grantToken` is the only endpoint in which the tenant ID cannot be obtained from the request path, query params, or ID token in the request body)
- Added additional unit tests for `grantToken` edge cases

Fixes https://github.com/firebase/firebase-tools/issues/4414. 

### Scenarios Tested

`npm test` passes

### Sample Commands

N/A

### CODE REVIEW

File: src/emulator/auth/state.ts [hunk 1]
```diff
@@ -872,9 +855,11 @@ export type Config = {
 };
 
 interface RefreshTokenRecord {
+  _AuthEmulatorRefreshTokenRecord: string;
```
Comment (yuchenshi) [inline src/emulator/auth/state.ts:858]: Drop `Record`

File: src/emulator/auth/state.ts [hunk 2]
```diff
@@ -922,6 +907,19 @@ interface TemporaryProofRecord {
...
 }
 
+export function encodeRefreshToken(refreshTokenRecord: RefreshTokenRecord): string {
+  return Buffer.from(JSON.stringify(refreshTokenRecord), "utf8").toString("base64");
+}
+
+export function decodeRefreshToken(refreshTokenString: string): RefreshTokenRecord {
+  try {
+    const json = Buffer.from(refreshTokenString, "base64").toString("utf8");
+    return JSON.parse(json) as RefreshTokenRecord;
```
Comment (yuchenshi) [inline src/emulator/auth/state.ts:917]: Let's validate the object has the `_AuthEmulatorRefreshToken` field to prevent misuse of other types of tokens, e.g. custom tokens, fake IDP tokens, mfa pending credentials

File: src/emulator/auth/state.ts (hunk 1 above)
Comment (lisajian) [reply src/emulator/auth/state.ts]: Done!

File: src/emulator/auth/state.ts (hunk 2 above)
Comment (lisajian) [reply src/emulator/auth/state.ts]: Cool, missed that assertion, thanks!

**Review by yuchenshi — APPROVED** [review]

**Comment by yuchenshi** [comment]
CHANGELOG please
